modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+
+	/// The tasks to be dispatched with data provideed by foreign state oracle
+	///
+	/// ActiveQuery: map QueryIndex => Option<VerifiableCallOF<T>>
+	#[pallet::storage]
+	#[pallet::getter(fn active_query)]
+	pub(super) type ActiveQuery<T: Config> = StorageMap<_, Identity, QueryIndex, VerifiableCallOf<T>, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Index key does not have an active query currently
+		NoMatchingCall,
+		/// Verifiable Call is too large
+		TooLargeVerifiableCall,
+		/// Query has expired
+		QueryExpired,
+		/// Query has not yet expired
+		QueryNotExpired,
+		/// Not account that requested query
+		NotQueryAccount,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Dispatch task with arbitrary data in origin.
+		///
+		/// - `query_index`: Unique index mapped to a particular query
+		/// - `data`: Aribtrary data to be injected into the call via origin
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn dispatch_task(origin: OriginFor<T>, query_index: QueryIndex, data: Vec<u8>) -> DispatchResult {
+			T::OracleOrigin::ensure_origin(origin)?;
+
+			let verifiable_call = ActiveQuery::<T>::take(query_index).ok_or(Error::<T>::NoMatchingCall)?;
+			// Check that query has not expired
+			ensure!(
+				verifiable_call.expiry > T::BlockNumberProvider::current_block_number(),
+				Error::<T>::QueryExpired
+			);
+			let result = verifiable_call.dispatchable_call.dispatch(RawOrigin { data }.into());
+
+			Self::deposit_event(Event::CallDispatched {
+				task_result: result.map(|_| ()).map_err(|e| e.error),
+			});
+			Ok(())
+		}
+
+		/// Remove Query that has expired so chain state does not bloat. This rewards the oracle
+		/// with half the query fee
+		///
+		/// - `query_index`: Unique index that is mapped to a particular query
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn remove_expired_call(origin: OriginFor<T>, query_index: QueryIndex) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let verifiable_call = ActiveQuery::<T>::take(query_index).ok_or(Error::<T>::NoMatchingCall)?;
+			// Make sure query is expired
+			ensure!(
+				verifiable_call.expiry <= T::BlockNumberProvider::current_block_number(),
+				Error::<T>::QueryNotExpired
+			);
+
+			// Gives half of reward for clearing expired query from storage
+			let reward: Balance = verifiable_call.oracle_reward.div(2u32);
+			T::Currency::transfer(&Self::account_id(), &who, reward, ExistenceRequirement::AllowDeath)?;
+
+			Self::deposit_event(Event::<T>::StaleQueryRemoved { query_index });
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	// Returns pallet account
+	pub fn account_id() -> T::AccountId {
+		T::PalletId::get().into_account()
+	}
+}
+
+impl<T: Config> ForeignChainStateQuery<T::AccountId, T::VerifiableTask> for Pallet<T> {
+	#[transactional]
+	fn query_task(who: &T::AccountId, length_bound: usize, dispatchable_call: T::VerifiableTask) -> DispatchResult {
```

Comment:
why the length bound is coming from caller? the call is also coming from caller so caller can just verify it.
I think it should be a constant in this pallet.
Or we could make it a `DepositPerByte` so the final deposit is `BaseFee + DepositPerByte * call_length` and then we don't need to limit the length

Reply:
Added a max call size limit, configurable via runtime.
I think foreign state oracle is mostly called by internal modules, so call size is unclear to the end user. They would have no idea the fee size.